cmd/dryrun/dryrun.go

```diff
@@ -0,0 +1,478 @@
+		log.Fatalf("Error starting: %v", err)
+	}
+}
+
+func execute(c *cli.Context) {
+	verbose := c.Bool(flagVerbose)
+	if verbose {
+		log.SetLevel(log.DebugLevel)
+	} else {
+		log.SetLevel(log.InfoLevel)
+	}
+
+	if err := dryRun(c); err != nil {
+		log.WithError(err).Errorf("dryrun failed")
+	}
+}
+
+func dryRun(c *cli.Context) error {
+	cluster, err := getStorageCluster(c)
+	if err != nil {
+		return err
+	}
+
+	apiextensions.SetInstance(apiextensions.New(fakeextclient.NewSimpleClientset()))
+	coreops.SetInstance(coreops.New(fakek8sclient.NewSimpleClientset()))
+
+	mockCtrl := gomock.NewController(nil)
+	defer mockCtrl.Finish()
+
+	dryRunDriver, err := storage.Get(pxutil.DriverName)
+	if err != nil {
+		log.Fatalf("Error getting dry-run Storage driver %v", err)
+	}
+
+	dryRunClient := testutil.FakeK8sClient()
+	if err = dryRunDriver.Init(dryRunClient, runtime.NewScheme(), record.NewFakeRecorder(100)); err != nil {
+		log.Fatalf("Error initializing Storage driver for dry run %v", err)
+	}
+
+	funcGetDir := func() string {
+		ex, err := os.Executable()
+		if err != nil {
+			log.Warningf("failed to get executable path")
+			return ""
+		}
+
+		return filepath.Dir(ex) + "/configs"
+	}
+	pxutil.SpecsBaseDir = funcGetDir
+
+	for _, comp := range component.GetAll() {
+		enabled := comp.IsEnabled(cluster)
+		if skipComponents[comp.Name()] {
+			// Only log a warning if component is enabled but not supported.
+			if enabled {
+				log.Warningf("component \"%s\": dry run is not supported yet, component enabled: %t.", comp.Name(), enabled)
+			}
+			continue
+		}
+
+		log.Infof("component \"%s\" enabled: %t", comp.Name(), enabled)
+		if enabled {
+			err := comp.Reconcile(cluster)
+			if ce, ok := err.(*component.Error); ok &&
+				ce.Code() == component.ErrCritical {
+				return err
+			} else if err != nil {
+				msg := fmt.Sprintf("Failed to setup %s. %v", comp.Name(), err)
+				log.Warningf(msg)
+			}
+		} else {
+			if err := comp.Delete(cluster); err != nil {
+				msg := fmt.Sprintf("Failed to cleanup %v. %v", comp.Name(), err)
+				log.Warningf(msg)
+			}
+		}
+	}
+
+	controller := storagecluster.Controller{Driver: dryRunDriver}
+	controller.SetKubernetesClient(dryRunClient)
+	objs, err := getAllObjects(controller, cluster)
+	if err != nil {
+		log.WithError(err).Error("failed to get all objects")
+	}
+
+	outputFile := c.String(flagOutputFile)
+	if outputFile == "" {
+		outputFile = defaultOutputFile
+	}
+
+	var sb strings.Builder
+	for _, obj := range objs {
+		log.Infof("Operator will deploy %s %s", obj.GetObjectKind().GroupVersionKind().Kind, obj.GetName())
+		bytes, err := yaml.Marshal(obj)
+		if err != nil {
+			return err
+		}
+
+		sb.WriteString(string(bytes))
+		sb.WriteString("---\n")
+	}
+
+	f, err := os.OpenFile(outputFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+	if err != nil {
+		log.WithError(err).Errorf("failed to open file %s", outputFile)
+		return err
+	}
+	defer f.Close()
+	_, err = f.WriteString(sb.String())
```

Comment:
would the sb too large to write everything at once?

Reply:
This is ok, in the backup, we use the same code to write to configmap, as when writing to configmap we need to write entirely. 

Regarding the migration validation, I am working on it, will be in the next PR.